main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala

```diff
@@ -37,13 +45,112 @@ object Stratifier extends Phase[TypedAst.Root, TypedAst.Root] {
     */
   def run(root: Root)(implicit flix: Flix): Validation[TypedAst.Root, CompilationError] = {
 
-    val b = System.nanoTime()
-
-    // TODO: Implement
+    val startTime = System.nanoTime()
 
-    val e = System.nanoTime() - b
+    val stratified = stratify(root.strata, root.tables.keys.toList)
+    val duration = System.nanoTime() - startTime
 
-    root.copy(time = root.time.copy(stratifier = e)).toSuccess
+    for {
+      stratified <- stratified
+    } yield root.copy(strata = stratified).copy(time = root.time.copy(stratifier = duration))
   }
 
+  /**
+    * Stratify the graph
+    */
+  def stratify(stratum: List[Stratum], syms: List[Symbol.TableSym]): Validation[List[Stratum], StratificationError] = {
+    // Implementing as described in Database and Knowledge - Base Systems Volume 1
+    // Ullman, Algorithm 3.5 p 133
+
+    /// Start by creating a mapping from predicates to stratum
+    var strataMap = syms.map(x => (x, 1)).toMap
+    val numRules = stratum.head.constraints.size
+
+    // We repeatedly examine the rules. We move the head predicate to
+    // the same strata that the body predicate is currently in if it is
+    // non negated, or one after the body, if it is. We repeat until
+    // there are no changes.
+    //
+    // If we create more strata than there are rules, then there is no
+    // stratification and we error out
```

Comment:
I found this algorithm in the Ullman book to do stratification. It implicitly checks for negative cycles, so I've removed the negative cycle check.